akka-cluster-typed/src/main/scala/akka/cluster/ddata/typed/javadsl/ReplicatorMessageAdapter.scala

```diff
@@ -0,0 +1,159 @@
+ * @param unexpectedAskTimeout The timeout to use for `ask` operations. This should be longer than
+ *                             the `timeout` given in [[Replicator.WriteConsistency]] and
+ *                             [[Replicator.ReadConsistency]]. The replicator will always send
+ *                             a reply within those timeouts so the `unexpectedAskTimeout` should
+ *                             not occur, but for cleanup in a failure situation it must still exist.
+ *                             If `askUpdate`, `askGet` or `askDelete` takes longer then this
+ *                             `unexpectedAskTimeout` a [[java.util.concurrent.TimeoutException]]
+ *                             will be thrown by the requesting actor and may be handled by supervision.
+ *
+ * @tparam A Message type of the requesting actor.
+ * @tparam B Type of the [[ReplicatedData]].
+ */
+class ReplicatorMessageAdapter[A, B <: ReplicatedData](
+    context: ActorContext[A],
+    replicator: ActorRef[Replicator.Command],
+    unexpectedAskTimeout: Duration) {
+
+  private implicit val askTimeout: Timeout = Timeout(unexpectedAskTimeout.asScala)
+
+  private var changedMessageAdapters: Map[Key[B], ActorRef[Replicator.Changed[B]]] = Map.empty
+
+  /**
+   * Subscribe to changes of the given `key`. The [[Replicator.Changed]] messages from
+   * the replicator are transformed to the message protocol of the requesting actor with
+   * the given `responseAdapter` function.
+   */
+  def subscribe(key: Key[B], responseAdapter: JFunction[Replicator.Changed[B], A]): Unit = {
+    // unsubscribe in case it's called more than once per key
+    unsubscribe(key)
+    changedMessageAdapters.get(key).foreach { subscriber =>
+      replicator ! Replicator.Unsubscribe(key, subscriber)
+    }
+    val replyTo: ActorRef[Replicator.Changed[B]] =
+      context.messageAdapter(classOf[Replicator.Changed[B]], responseAdapter)
+    changedMessageAdapters = changedMessageAdapters.updated(key, replyTo)
+    replicator ! Replicator.Subscribe(key, replyTo)
+  }
+
+  /**
+   * Unsubscribe from a previous subscription of a given `key`.
+   * @see [[ReplicatorMessageAdapter.subscribe]]
+   */
+  def unsubscribe(key: Key[B]): Unit = {
+    changedMessageAdapters.get(key).foreach { subscriber =>
+      replicator ! Replicator.Unsubscribe(key, subscriber)
+    }
+  }
+
+  /**
+   * Send a [[Replicator.Update]] request to the replicator. The [[Replicator.UpdateResponse]]
+   * message is transformed to the message protocol of the requesting actor with the given
+   * `responseAdapter` function.
+   *
+   * Note that `createRequest` is a function that creates the `Update` message from the provided
+   * `ActorRef[UpdateResponse]` that the the replicator will send the response message back through.
+   * Use that `ActorRef[UpdateResponse]` as the `replyTo` parameter in the `Update` message.
+   */
+  def askUpdate(
```

Comment:
E.g

```scala
      replicatorAdapter.upate(
                  key,
                  curr -> curr.increment(node, 1),
                  InternalUpdateResponse::new);
```

And the impl

```scala
  def upate(
        key: Key[B], 
        update: JFunction[B, B],
        responseAdapter: JFunction[Replicator.UpdateResponse[B], A]): Unit = {
    context.asScala.ask[Replicator.Update[B], Replicator.UpdateResponse[B]](replicator)(askReplyTo =>
      new Replicator.Update[B](key, initial, Replicator.writeLocal, askReplyTo, update)) {
      case Success(value) => responseAdapter(value)
      case Failure(ex)    => throw ex // unexpected ask timeout
    }
  }
```

This hides the actorref => functions that can be confusing but users are exposed to anyway when using ask

Reply:
But with the consistency defaulted on construction and maybe an overload to specify per request

Reply:
I was considering that but I think it's still nice to have the messages as parameters to these, because otherwise we will probably need a few overloaded variants and it will be more difficult to maintain if adding something more to them.